src/rules/arrayTypeRule.ts

```diff
@@ -51,83 +53,116 @@ export class Rule extends Lint.Rules.AbstractRule {
+        }
+        return ts.forEachChild(node, cb);
+    });
+
+    function checkArrayType(node: ts.ArrayTypeNode): void {
+        const { elementType, parent } = node;
+        if (option === "array" || option === "array-simple" && isSimpleType(elementType)) {
+            return;
         }
 
-        super.visitArrayType(node);
+        const failureString = option === "generic" ? Rule.FAILURE_STRING_GENERIC : Rule.FAILURE_STRING_GENERIC_SIMPLE;
+        const parens = elementType.kind === ts.SyntaxKind.ParenthesizedType ? 1 : 0;
+        // Add a space if the type is preceded by 'as' and the node has no leading whitespace
+        const space = !parens && parent!.kind === ts.SyntaxKind.AsExpression && node.getStart() === node.getFullStart();
+        const fix = ctx.createFix(
+            new Lint.Replacement(elementType.getStart(), parens, (space ? " " : "") + "Array<"),
+            // Delete the square brackets and replace with an angle bracket
+            Lint.Replacement.replaceFromTo(elementType.getEnd() - parens, node.getEnd(), ">"));
+        ctx.addFailureAtNode(node, failureString, fix);
     }
 
-    public visitTypeReference(node: ts.TypeReferenceNode) {
-        const typeName = node.typeName.getText();
-        if (typeName === "Array" && (this.hasOption(OPTION_ARRAY) || this.hasOption(OPTION_ARRAY_SIMPLE))) {
-            const failureString = this.hasOption(OPTION_ARRAY) ? Rule.FAILURE_STRING_ARRAY : Rule.FAILURE_STRING_ARRAY_SIMPLE;
-            const typeArgs = node.typeArguments;
-            if (!typeArgs || typeArgs.length === 0) {
-                // Create an 'any' array
-                const fix = this.createFix(this.createReplacement(node.getStart(), node.getWidth(), "any[]"));
-                this.addFailureAtNode(node, failureString, fix);
-            } else if (typeArgs && typeArgs.length === 1 && (!this.hasOption(OPTION_ARRAY_SIMPLE) || this.isSimpleType(typeArgs[0]))) {
-                const type = typeArgs[0];
-                const typeStart = type.getStart();
-                const typeEnd = type.getEnd();
-                const parens = type.kind === ts.SyntaxKind.UnionType ||
-                    type.kind === ts.SyntaxKind.FunctionType || type.kind === ts.SyntaxKind.IntersectionType;
-                const fix = this.createFix(
-                    // Delete Array and the first angle bracket
-                    this.createReplacement(node.getStart(), typeStart - node.getStart(), parens ? "(" : ""),
-                    // Delete the last angle bracket and replace with square brackets
-                    this.createReplacement(typeEnd, node.getEnd() - typeEnd, (parens ? ")" : "") + "[]"),
-                );
-                this.addFailureAtNode(node, failureString, fix);
-            }
+    function checkTypeReference(node: ts.TypeReferenceNode): void {
+        const { typeName, typeArguments } = node;
+
+        if (option === "generic" || !isArrayIdentifier(typeName)) {
```

Comment:
to be consistent. you should use the constants. e.g. `OPTION_GENERIC` here and in all other comparisons below